[Experimental] Add Product Reviews Pagination block

dinhtungdu opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR adds Product Reviews Pagination and its inner blocks. I decided to add them all in a PR for easier testing. New blocks are mimic from Gutenberg with some modification to fit with WooCommerce.

Closes https://github.com/woocommerce/woocommerce/issues/57124

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

<img width="1514" alt="image" src="https://github.com/user-attachments/assets/54b35f00-a20b-481a-b7a0-03a9a1eaadec" />

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Enable Comment Pagination in WP Dashboard > Settings > Discussion. Set the `Top level comments per page` setting to a small value if you don't have many reviews on a product.
2. Add the `Blockified Product Details` block to Single Product template.
3. Expand the Reviews item, see new Product Reviews Pagination block.
4. Adjust block settings, see editor preview react accordingly.
5. Save and test on the front end with a product that have multiple reviews, verify the pagination works and customization is applied.
6. Test other product with no review, see block isn't rendered.
<!-- End testing instructions -->

dinhtungdu committed:
feat: product reviews pagination block

dinhtungdu committed:
feat: reviews pagination next

dinhtungdu committed:
feat: reviews pagination previous

dinhtungdu committed:
feat: reviews-pagination-numbers block

dinhtungdu committed:
chore: group accordion blocks registration

dinhtungdu committed:
fix: bail frontend script registration for product details block

dinhtungdu committed:
chore: changelog

dinhtungdu committed:
fix: change default justify content

dinhtungdu committed:
fix: use parent for pagination inner blocks

dinhtungdu committed:
Revert "fix: change default justify content"

This reverts commit 76591ff51ba5c5c0365b427b3ac6c3d575d13012.

dinhtungdu committed:
fix: pagination arrows

dinhtungdu committed:
fix: use comment word when mention Discussion settings

dinhtungdu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination/edit.tsx:
```diff
@@ -0,0 +1,100 @@
+}: Props ) {
+	const hasNextPreviousBlocks = useSelect(
+		( select ) => {
+			const { getBlocks } = select( blockEditorStore );
```
The type for `@wordpress/block-editor`'s store is outdated. It's typed properly in more recent version. Because of that, I decided this as is without adding any comment. The warning will be solved automatically when we update the type packages.

mikachan commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination-previous/index.tsx:
```diff
@@ -0,0 +1,26 @@
+import edit from './edit';
+
+const { name } = metadata;
+export { metadata, name };
```
```suggestion
```

It looks like we can remove `name` and `metadata` from here, as I don't think they're used anywhere.

mikachan commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination-numbers/index.tsx:
```diff
@@ -0,0 +1,20 @@
+import edit from './edit';
+
+const { name } = metadata;
+export { metadata, name };
```
```suggestion
```

Similar here - looks like we can remove `name` and `metadata`.

mikachan commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination-numbers/index.tsx:
```diff
@@ -0,0 +1,20 @@
+const { name } = metadata;
+export { metadata, name };
+
+registerBlockType( metadata, {
```
```suggestion
// @ts-expect-error metadata is not typed.
registerBlockType( metadata, {
```

This is missing the usual TS comment for fixing the missing types for `registerBlockType`.

mikachan commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination-next/index.tsx:
```diff
@@ -0,0 +1,26 @@
+import metadata from './block.json';
+import edit from './edit';
+
+const { name } = metadata;
+export { metadata, name };
```
```suggestion
```

Same here again I think - we can remove `name` and `metadata`.

mikachan commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination/edit.tsx:
```diff
@@ -0,0 +1,100 @@
+	} );
+
+	// Get the Discussion settings
+	const pageComments = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		const { __experimentalDiscussionSettings } = getSettings();
+		return __experimentalDiscussionSettings?.pageComments;
+	}, [] );
```
Would it be better to use `getEditorSettings` from `core/editor` here to get the discussion settings? I see this is being used in some of the product fields blocks to get settings.

dinhtungdu replied:
Is the discussion setting included there? Looks like it isn't: https://github.com/WordPress/wordpress-develop/blob/001a86a7ca833535533e59dc59babddc50fe2afa/src/wp-includes/block-editor.php#L155-L237

mikachan replied:
I see it returned when I test with `getEditorSettings` on this branch. Here's a console log of (part of) the settings object I'm seeing:

<img width="580" alt="image" src="https://github.com/user-attachments/assets/70c30904-963e-4eeb-9a43-6f002f60ad69" />

Looks like it's included here: https://github.com/WordPress/wordpress-develop/blob/0aa29fd973213146d7b7875d17d3729990eb66a6/src/wp-includes/block-editor.php#L629.

Here's what I used for testing if it's helpful:

```
	// Get the Discussion settings
	const pageComments = useSelect( ( select ) => {
		const { getEditorSettings } = select( 'core/editor' ) as {
			getEditorSettings: () => Record< string, unknown >;
		};
		const settings = getEditorSettings();
		console.log( 'settings', settings );
		// eslint-disable-next-line @typescript-eslint/naming-convention
		const { __experimentalDiscussionSettings } = settings as {
			// eslint-disable-next-line @typescript-eslint/naming-convention
			__experimentalDiscussionSettings?: { pageComments?: boolean };
		};
		return __experimentalDiscussionSettings?.pageComments;
	}, [] );
```

mikachan commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-pagination/edit.tsx:
```diff
@@ -0,0 +1,100 @@
+			 * Product Reviews Pagination Next or Product Reviews Pagination Previous
+			 * block exists.
+			 */
+			return innerBlocks?.find( ( innerBlock ) => {
```
```suggestion
			return innerBlocks?.find( ( innerBlock: { name: string } ) => {
```

I think we can fix the type error here like this.

dinhtungdu replied:
As I commented in https://github.com/woocommerce/woocommerce/pull/57604/files#r2067975246, the reason for the error is the incorrect and outdated type of `@wordpress/block-editor`'s store. It's typed correctly in the recent version, so when we update the type package, the error will be inferred correctly. That's why I leave it as is. But maybe a comment explain that will help?

mikachan replied:
Ah I see, I thought that comment was just referring to the type error for `getBlocks`. Yep OK, I think it's fine to leave it as is then.

mikachan reviewed (commented):
For some reason, I'm seeing block support errors for the pagination blocks when I test this branch:

<img width="670" alt="image" src="https://github.com/user-attachments/assets/378dc30e-01a9-4fc8-95e4-dce9c23442f7" />

I'm guessing this is maybe a problem with the way I'm building the project or something 😅 

For the time-being, I've left some inline comments, but I don't think any are blocking. Code is looking good!

dinhtungdu committed:
Apply suggestions from code review

Co-authored-by: Sarah Norris <1645628+mikachan@users.noreply.github.com>

dinhtungdu commented:
> For some reason, I'm seeing block support errors for the pagination blocks when I test this branch:

@mikachan I got the same issue working on this PR too. For an unknown reason, blocks aren't built properly, the `block.json` file isn't copied over. I have to delete the `.wireit` cache to get it building again. You can find the `.wireit` cache directory in `plugins/woocommerce/` and `plugins/woocommerce/client/blocks/`. Let me know if it works for you.

dinhtungdu committed:
chore: lint

dinhtungdu committed:
fix: block build for inner blocks

If we define the `parent` attribute in block.json, the metadata file
will be copied to `inner-blocks` directory in the build directory, which
requires block to extend the AbstractInnerBlock. We don't want to use
that class because it was made for frontend rendered blocks.

dinhtungdu commented:
@mikachan [`9e3771a` (#57604)](https://github.com/woocommerce/woocommerce/pull/57604/commits/9e3771ab435af543543d897166d257d994e532e2) should fix the build issue.

mikachan reviewed (approved):
> [9e3771a (#57604)](https://github.com/woocommerce/woocommerce/pull/57604/commits/9e3771ab435af543543d897166d257d994e532e2) should fix the build issue.

Yes! All blocks are working well for me now:

Editor:
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/7340f7f6-332d-4afc-b031-843f39f362aa" />

Front end:
<img width="709" alt="image" src="https://github.com/user-attachments/assets/2e3f258d-b2d7-4e0b-a477-a30e3f6d287c" />

dinhtungdu committed:
chore: add todo and comment about outdated types

dinhtungdu commented:
@mikachan in [`9d48ce4` (#57604)](https://github.com/woocommerce/woocommerce/pull/57604/commits/9d48ce496b190205c54b7f6d819e771d61b84a91), I added some comments and a todo about the outdated types we're using for `@wordpress/block-editor`. I decided to not use `core/editor` to get the discussion setting because we're still to cast the `getSettings` type. Once we upgrade the type package, we can remove the comment added in this commit.
